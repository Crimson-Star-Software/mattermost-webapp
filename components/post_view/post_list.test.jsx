// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import LoadingScreen from 'components/loading_screen';
import DateSeparator from 'components/post_view/date_separator';
import Post from 'components/post_view/post';

import PostList from './post_list';

describe('PostList', () => {
    const baseProps = {
        channel: {id: 'channel'},
        currentUserId: '',
        focusedPostId: '',
        fullWidth: true,
        lastViewedAt: 0,
        posts: [
            {id: 'post1', message: 'Hello, World 1!', create_at: new Date(2019, 2, 4, 12).getTime()},
            {id: 'post2', message: 'Hello, World 2!', create_at: new Date(2019, 2, 4, 11).getTime()},
            {id: 'post3', message: 'Hello, World 3!', create_at: new Date(2019, 2, 4, 10).getTime()},
        ],
        postVisibility: 10,
        actions: {
            checkAndSetMobileView: jest.fn(),
            increasePostVisibility: jest.fn(),
            loadInitialPosts: jest.fn(() => ({posts: {posts: {}, order: []}, hasMoreBefore: false})),
        },
    };

    test('should render loading screen while loading posts', () => {
        const props = {
            ...baseProps,
            posts: [],
        };

        const wrapper = shallow(<PostList {...props}/>);

        expect(wrapper.find(LoadingScreen).exists()).toBe(true);

        wrapper.setState({isDoingInitialLoad: false});

        expect(wrapper.find(LoadingScreen).exists()).toBe(false);
    });

    describe('posts', () => {
        test('should render correctly', () => {
            const wrapper = shallow(<PostList {...baseProps}/>);

            wrapper.setState({isDoingInitialLoad: false});

            expect(wrapper.find(Post)).toHaveLength(3);
        });

        test('should limit the number based on postVisibility', () => {
            const props = {
                ...baseProps,
                postVisibility: 2,
            };

            const wrapper = shallow(<PostList {...props}/>);

            wrapper.setState({isDoingInitialLoad: false});

            expect(wrapper.find(Post)).toHaveLength(2);
        });
    });

    describe('date separators', () => {
        test('should render above posts made all on the same day', () => {
            const wrapper = shallow(<PostList {...baseProps}/>);

            wrapper.setState({
                atEnd: true,
                isDoingInitialLoad: false,
            });

            expect(wrapper.find(DateSeparator)).toHaveLength(1);
            expect(wrapper.find('.post-list__content').children()).toMatchSnapshot();
        });

        test('should render between posts made on different days', () => {
            const props = {
                ...baseProps,
                posts: [
                    {id: 'post1', message: 'Hello, World 1!', create_at: new Date(2019, 2, 4, 12).getTime()},
                    {id: 'post2', message: 'Hello, World 2!', create_at: new Date(2019, 2, 4, 11).getTime()},
                    {id: 'post3', message: 'Hello, World 3!', create_at: new Date(2019, 2, 3, 12).getTime()},
                ],
            };

            const wrapper = shallow(<PostList {...props}/>);

            wrapper.setState({
                atEnd: true,
                isDoingInitialLoad: false,
            });

            expect(wrapper.find(DateSeparator)).toHaveLength(2);
            expect(wrapper.find('.post-list__content').children()).toMatchSnapshot();
        });
    });

    describe('new message line', () => {
        test('should not render on first read', () => {
            const props = {
                ...baseProps,
                lastViewedAt: 0,
            };

            const wrapper = shallow(<PostList {...props}/>);

            wrapper.setState({
                atEnd: true,
                isDoingInitialLoad: false,
            });

            expect(wrapper.find('.new-separator').exists()).toBe(false);
        });

        test('should render new message line between posts', () => {
            const props = {
                ...baseProps,
                lastViewedAt: new Date(2019, 2, 3, 13).getTime(),
            };

            const wrapper = shallow(<PostList {...props}/>);

            wrapper.setState({
                atEnd: true,
                isDoingInitialLoad: false,
            });

            expect(wrapper.find('.new-separator').exists()).toBe(true);
            expect(wrapper.find('.post-list__content').children()).toMatchSnapshot();
        });

        test('should not render when all posts have been read', () => {
            const props = {
                ...baseProps,
                lastViewedAt: baseProps.posts[0].create_at,
            };

            const wrapper = shallow(<PostList {...props}/>);

            wrapper.setState({
                atEnd: true,
                isDoingInitialLoad: false,
            });

            expect(wrapper.find('.new-separator').exists()).toBe(false);
        });

        test('should not render on first read', () => {
            const props = {
                ...baseProps,
                lastViewedAt: 0,
            };

            const wrapper = shallow(<PostList {...props}/>);

            wrapper.setState({
                atEnd: true,
                isDoingInitialLoad: false,
            });

            expect(wrapper.find('.new-separator').exists()).toBe(false);
        });
    });
});
