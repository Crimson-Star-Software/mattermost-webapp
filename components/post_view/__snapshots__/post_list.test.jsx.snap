// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList date separators should render above posts made all on the same day 1`] = `
Array [
  <Connect(ChannelIntroMessage)
    channel={
      Object {
        "id": "channel",
      }
    }
    fullWidth={true}
  />,
  <DateSeparator
    date={1551711600000}
    key="1551711600000"
  />,
  <Connect(Post)
    getPostList={[Function]}
    key="post post3"
    lastPostCount={-1}
    post={
      Object {
        "create_at": 1551711600000,
        "id": "post3",
        "message": "Hello, World 3!",
      }
    }
  />,
  <Connect(Post)
    getPostList={[Function]}
    key="post post2"
    lastPostCount={-1}
    post={
      Object {
        "create_at": 1551715200000,
        "id": "post2",
        "message": "Hello, World 2!",
      }
    }
  />,
  <Connect(Post)
    getPostList={[Function]}
    key="post post1"
    lastPostCount={-1}
    post={
      Object {
        "create_at": 1551718800000,
        "id": "post1",
        "message": "Hello, World 1!",
      }
    }
  />,
]
`;

exports[`PostList date separators should render between posts made on different days 1`] = `
Array [
  <Connect(ChannelIntroMessage)
    channel={
      Object {
        "id": "channel",
      }
    }
    fullWidth={true}
  />,
  <DateSeparator
    date={1551632400000}
    key="1551632400000"
  />,
  <Connect(Post)
    getPostList={[Function]}
    key="post post3"
    lastPostCount={-1}
    post={
      Object {
        "create_at": 1551632400000,
        "id": "post3",
        "message": "Hello, World 3!",
      }
    }
  />,
  <DateSeparator
    date={1551715200000}
    key="1551715200000"
  />,
  <Connect(Post)
    getPostList={[Function]}
    key="post post2"
    lastPostCount={-1}
    post={
      Object {
        "create_at": 1551715200000,
        "id": "post2",
        "message": "Hello, World 2!",
      }
    }
  />,
  <Connect(Post)
    getPostList={[Function]}
    key="post post1"
    lastPostCount={-1}
    post={
      Object {
        "create_at": 1551718800000,
        "id": "post1",
        "message": "Hello, World 1!",
      }
    }
  />,
]
`;

exports[`PostList new message line should render new message line between posts 1`] = `
Array [
  <Connect(ChannelIntroMessage)
    channel={
      Object {
        "id": "channel",
      }
    }
    fullWidth={true}
  />,
  <DateSeparator
    date={1551711600000}
    key="1551711600000"
  />,
  <div
    className="new-separator"
    id="new_message_post3"
    key="unviewed"
  >
    <hr
      className="separator__hr"
    />
    <div
      className="separator__text"
    >
      <FormattedMessage
        defaultMessage="New Messages"
        id="posts_view.newMsg"
        values={Object {}}
      />
    </div>
  </div>,
  <Connect(Post)
    getPostList={[Function]}
    key="post post3"
    lastPostCount={-1}
    post={
      Object {
        "create_at": 1551711600000,
        "id": "post3",
        "message": "Hello, World 3!",
      }
    }
  />,
  <Connect(Post)
    getPostList={[Function]}
    key="post post2"
    lastPostCount={-1}
    post={
      Object {
        "create_at": 1551715200000,
        "id": "post2",
        "message": "Hello, World 2!",
      }
    }
  />,
  <Connect(Post)
    getPostList={[Function]}
    key="post post1"
    lastPostCount={-1}
    post={
      Object {
        "create_at": 1551718800000,
        "id": "post1",
        "message": "Hello, World 1!",
      }
    }
  />,
]
`;
